File: SproInstall/SocketPro/tutorial/CSharp/SampleThree/Server/TThreeImpl.cs
/* **** including all of defines, service id(s) and request id(s) ***** */
using System;
using SocketProAdapter;
using SocketProAdapter.ServerSide;
using USOCKETLib; //you may need it for accessing various constants
using SampleThreeShared;
using System.Collections;

//server implementation for service CTThree
public class CTThreePeer : CClientPeer
{
	protected override void OnSwitchFrom(int nServiceID)
	{
        m_TThreeSvs = (CTThreeSvs)GetBaseService();
        Console.WriteLine("Socket is switched for the service CTThreeSvs");
	}

	protected override void OnReleaseResource(bool bClosing, int nInfo)
	{
		if(bClosing)
		{
            Console.WriteLine("Socket is closed with error code = " + nInfo);
		}
		else
		{
            Console.WriteLine("Socket is going to be switched to new service with service id = " + nInfo);
		}
        m_Stack.Clear();
	}

	protected void GetOneItem(out CTestItem GetOneItemRtn)
	{
        if (m_TThreeSvs.m_Stack.Count > 0)
            GetOneItemRtn = (CTestItem)m_TThreeSvs.m_Stack.Pop();
        else
            GetOneItemRtn = null;
	}

	protected void SendOneItem(CTestItem Item)
	{
        m_TThreeSvs.m_Stack.Push(Item);
	}

	protected void GetManyItems()
	{
        int nRtn = 0;
        m_UQueue.SetSize(0);
        while (m_Stack.Count > 0)
        {
            //a client may either shut down the socket connection or call IUSocket::Cancel
            if (nRtn == SOCKET_NOT_FOUND || nRtn == REQUEST_CANCELED)
                break;
            CTestItem Item = (CTestItem)m_Stack.Pop();
            m_UQueue.Push(Item);
           
            //20 kbytes per batch at least
            //also shouldn't be too large. 
            //If the size is too large, it will cost more memory resource and reduce conccurency if online compressing is enabled.
            //for an opimal value, you'd better test it by yourself
            if (m_UQueue.GetSize() > 20480)
            {
                nRtn = SendResult(TThreeConst.idGetBatchItemsCTThree, m_UQueue);
                m_UQueue.SetSize(0);
            }
        }
        if (nRtn == SOCKET_NOT_FOUND || nRtn == REQUEST_CANCELED)
        {

        }
        else if (m_UQueue.GetSize() > sizeof(int))
        {
            nRtn = SendResult(TThreeConst.idGetBatchItemsCTThree, m_UQueue);
        }
	}

	protected void SendManyItems()
	{
	    while(m_Stack.Count > 0)
	    {
            m_TThreeSvs.m_Stack.Push(m_Stack.Pop());
	    }
	}

    protected void SendBatchItems()
    {
        while(m_UQueue.GetSize() > 0)
	    {
		    CTestItem Item = new CTestItem();
            m_UQueue.Pop(out Item);
            m_Stack.Push(Item);
	    }
    }

	protected override void OnFastRequestArrive(short sRequestID, int nLen)
	{
		switch(sRequestID)
		{
		case TThreeConst.idGetOneItemCTThree:
            M_I0_R1<CTestItem>(GetOneItem);
			break;
		case TThreeConst.idSendOneItemCTThree:
            M_I1_R0<CTestItem>(SendOneItem);
			break;
		case TThreeConst.idSendManyItemsCTThree:
            M_I0_R0(SendManyItems);
			break;
		default:
			break;
		}
	}

    private int RetrieveCount()
    {
        int nCount = 0;
        m_UQueue.Pop(out nCount);
        return nCount;
    }

    protected override void OnDispatchingSlowRequest(short sRequestID)
    {
        if(sRequestID == TThreeConst.idGetManyItemsCTThree) 
	    {
            int nCount = RetrieveCount();
            m_Stack.Clear();
            while (m_TThreeSvs.m_Stack.Count > 0 && nCount > 0)
            {
                m_Stack.Push(m_TThreeSvs.m_Stack.Pop());
                nCount--;
            }
	    }
    }

	protected override int OnSlowRequestArrive(short sRequestID, int nLen)
	{
		switch(sRequestID)
		{
                
		case TThreeConst.idGetManyItemsCTThree:
            M_I0_R0(GetManyItems);
			break;
        case TThreeConst.idSendBatchItemsCTThree:
            M_I0_R0(SendBatchItems);
            break;
		default:
			break;
		}
		return 0;
	}
    private Stack m_Stack = new Stack();
    private CTThreeSvs m_TThreeSvs;
}

public class CTThreeSvs : CSocketProService<CTThreePeer>
{
    internal Stack m_Stack = new Stack();
}

File: SproInstall/SocketPro/tutorial/VBNet/SampleThree/Server/TThreeImpl.vb
' **** including all of defines, service id(s) and request id(s) ***** 
Imports SocketProAdapter
Imports SocketProAdapter.ServerSide
Imports USOCKETLib 'you may need it for accessing various constants
Imports SampleThreeShared
Imports System.Collections

'server implementation for service CTThree
Public Class CTThreePeer
	Inherits CClientPeer
	Protected Overrides Sub OnSwitchFrom(ByVal nServiceID As Integer)
		m_TThreeSvs = CType(GetBaseService(), CTThreeSvs)
		Console.WriteLine("Socket is switched for the service CTThreeSvs")
	End Sub

	Protected Overrides Sub OnReleaseResource(ByVal bClosing As Boolean, ByVal nInfo As Integer)
		If bClosing Then
			Console.WriteLine("Socket is closed with error code = " & nInfo)
		Else
			Console.WriteLine("Socket is going to be switched to new service with service id = " & nInfo)
		End If
		m_Stack.Clear()
	End Sub

	Protected Sub GetOneItem(ByRef GetOneItemRtn As CTestItem)
		If m_TThreeSvs.m_Stack.Count > 0 Then
			GetOneItemRtn = CType(m_TThreeSvs.m_Stack.Pop(), CTestItem)
		Else
            GetOneItemRtn = Nothing
		End If
	End Sub

	Protected Sub SendOneItem(ByVal Item As CTestItem)
		m_TThreeSvs.m_Stack.Push(Item)
	End Sub

	Protected Sub GetManyItems()
		Dim nRtn As Integer = 0
		m_UQueue.SetSize(0)
		Do While m_Stack.Count > 0
			'a client may either shut down the socket connection or call IUSocket::Cancel
			If nRtn = SOCKET_NOT_FOUND OrElse nRtn = REQUEST_CANCELED Then
				Exit Do
			End If
			Dim Item As CTestItem = CType(m_Stack.Pop(), CTestItem)
            m_UQueue.Push(Item)

			'20 kbytes per batch at least
			'also shouldn't be too large. 
			'If the size is too large, it will cost more memory resource and reduce conccurency if online compressing is enabled.
			'for an opimal value, you'd better test it by yourself
			If m_UQueue.GetSize() > 20480 Then
				nRtn = SendResult(TThreeConst.idGetBatchItemsCTThree, m_UQueue)
				m_UQueue.SetSize(0)
			End If
		Loop
		If nRtn = SOCKET_NOT_FOUND OrElse nRtn = REQUEST_CANCELED Then

		ElseIf m_UQueue.GetSize() > Len(New Integer) Then
			nRtn = SendResult(TThreeConst.idGetBatchItemsCTThree, m_UQueue)
		End If
	End Sub

	Protected Sub SendManyItems()
		Do While m_Stack.Count > 0
			m_TThreeSvs.m_Stack.Push(m_Stack.Pop())
		Loop
	End Sub

	Protected Sub SendBatchItems()
		Do While m_UQueue.GetSize() > 0
            Dim Item As CTestItem = Nothing
            m_UQueue.Pop(Item)
			m_Stack.Push(Item)
		Loop
	End Sub

	Protected Overrides Sub OnFastRequestArrive(ByVal sRequestID As Short, ByVal nLen As Integer)
		Select Case sRequestID
		Case TThreeConst.idGetOneItemCTThree
			M_I0_R1(Of CTestItem)(AddressOf GetOneItem)
		Case TThreeConst.idSendOneItemCTThree
			M_I1_R0(Of CTestItem)(AddressOf SendOneItem)
		Case TThreeConst.idSendManyItemsCTThree
			M_I0_R0(AddressOf SendManyItems)
		Case Else
		End Select
	End Sub

	Private Function RetrieveCount() As Integer
		Dim nCount As Integer = 0
		m_UQueue.Pop(nCount)
		Return nCount
	End Function

	Protected Overrides Sub OnDispatchingSlowRequest(ByVal sRequestID As Short)
		If sRequestID = TThreeConst.idGetManyItemsCTThree Then
			Dim nCount As Integer = RetrieveCount()
			m_Stack.Clear()
			Do While m_TThreeSvs.m_Stack.Count > 0 AndAlso nCount > 0
				m_Stack.Push(m_TThreeSvs.m_Stack.Pop())
				nCount -= 1
			Loop
		End If
	End Sub

	Protected Overrides Function OnSlowRequestArrive(ByVal sRequestID As Short, ByVal nLen As Integer) As Integer
		Select Case sRequestID

		Case TThreeConst.idGetManyItemsCTThree
			M_I0_R0(AddressOf GetManyItems)
		Case TThreeConst.idSendBatchItemsCTThree
			M_I0_R0(AddressOf SendBatchItems)
		Case Else
		End Select
		Return 0
	End Function
	Private m_Stack As New Stack()
	Private m_TThreeSvs As CTThreeSvs
End Class

Public Class CTThreeSvs
    Inherits CSocketProService(Of CTThreePeer)

	Friend m_Stack As New Stack()
End Class

